### PR TITLE
Support rotatating the JWT shared secret

### PR DESCRIPTION
**Pull request by jakelandis**
This commit adds support to reload the JWT shared secret. 
Notably this commit also includes support for a rotatable secret which includes
support for a configurable grace period where the elder value (after rotation)
is still accessible. This allows a time bound leniency where both values are 
valid at the same time to help mitigate tightly coupled systems rotations. 
The rotatable secret currently only supports checking if it is set or it matches 
an external secret. However, future updates will accept a function that
can be be used as input to a 3rd party system that can try the current secret
but automatically fall back to the prior secret if that fails during the grace period. 
The implementation of rotatable secret uses a StampedLock with optimistic 
reads to help ensure minimal performance impact for the reading and expiry 
of the secret.

### CODE REVIEW

**Comment by jakelandis** [comment]
doc preview here : https://elasticsearch_99278.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/security-settings.html

File: server/src/test/java/org/elasticsearch/common/settings/RotatableSecretTests.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        });
+
+        // start writer thread
+        Thread t1 = new Thread(() -> rotatableSecret.rotate(secret2, mockGracePeriod));
+        t1.start();
+        assertBusy(() -> assertEquals(Thread.State.WAITING, t1.getState())); // waiting on countdown latch, holds write lock
+        assertTrue(rotatableSecret.isWriteLocked());
+
+        // start reader threads
+        int readers = randomIntBetween(0, 16);
```
Comment (adamkasztenny) [inline server/src/test/java/org/elasticsearch/common/settings/RotatableSecretTests.java:115]: Should we have at least 1 reader here so the loop after this is guaranteed to run?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/jwt/JwtRealmSettings.java [hunk 2]
```diff
@@ -172,7 +173,8 @@ private static Set<Setting.AffixSetting<?>> getNonSecureSettings() {
                 CLAIMS_MAIL.getPattern(),
                 CLAIMS_NAME.getClaim(),
                 CLAIMS_NAME.getPattern(),
-                POPULATE_USER_METADATA
+                POPULATE_USER_METADATA,
+                CLIENT_AUTHENTICATION_SHARED_SECRET_ROTATION_GRACE
```
Comment (adamkasztenny) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/jwt/JwtRealmSettings.java:177]: Should this be `CLIENT_AUTHENTICATION_SHARED_SECRET_ROTATION_GRACE_PERIOD`?

**Review by AthenaEryma — COMMENTED** [review]
Mostly looks good, nicely done.

File: server/src/test/java/org/elasticsearch/common/settings/RotatableSecretTests.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class RotatableSecretTests extends ESTestCase {
+
+    private final SecureString secret1 = new SecureString(randomAlphaOfLength(10));
+    private final SecureString secret2 = new SecureString(randomAlphaOfLength(10));
+    private final SecureString secret3 = new SecureString(randomAlphaOfLength(10));
+
+    public void testBasicRotation() throws Exception {
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/test/java/org/elasticsearch/common/settings/RotatableSecretTests.java:30]: added a new test for this.  helped to discover a missing usecase - starting with null then rotating to value.

**Review by AthenaEryma — APPROVED** [review]
LGTM, thanks!

**Pull request merged**